# Project: WeRateDogs Twitter Archive

## Table of Contents
<ul>
    <li><a href="#intro">Introduction</a></li>
    <li><a href="#wrangle">Data Wrangling</a></li>
    <li><a href="#analyze">Data Analysis</a></li>
    <li><a href="#visualize">Data Visualization</a></li>
    <li><a href="#conclusion">Conclusion</a></li>
</ul>

## Introduction
<a id='intro'></a>

> text

> text

In [129]:
# import libraries
import glob
import json
import os
import pandas as pd
import requests
import tweepy

## Data Wrangling
<a id='wrangle'></a>

### Gathering

In [2]:
# read in twitter-archive-enhanced file
df1 = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# get response object of info from url
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

In [4]:
# create tsv file for image predictions if it doesn't already exist
if not os.path.exists('image-predictions.tsv'):
    with open(os.path.join(url.split('/')[-1]), mode = 'wb') as file:
        file.write(response.content)

In [5]:
# read image predictions into pandas DataFrame
df2 = pd.read_csv('image-predictions.tsv', delimiter='\t', encoding='utf-8')

In [6]:
# use key, token, and secrets to access Twitter API
# REMOVE KEYS WHEN SUBMITTING!!!


# use tweepy to access api
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

> Althought the following section could have been done simpler by pulling the three columns of information from the API directly into the df3_list and then converting it into a DataFrame, it was stated in the project details that the tweets need to be stored in .txt files and then read into the DataFrame.

In [7]:
# make folder to store tweet info if it doesn't already exist
folder_name = 'tweet_infos'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

In [8]:
# create error list to catch tweet_ids where the tweet_info doesn't get pulled
error_list = []
# go through tweet_ids in df1 and saving tweet_info as txt files in tweet_infos folder
for tweet_id in df1.tweet_id:
    try:
        tweet_info = json.dumps(api.get_status(str(tweet_id), tweet_mode='extended')._json)
        with open(os.path.join(folder_name, '{}{}'.format(tweet_id,'.txt')), mode='w') as file:
            file.write(tweet_info)
            print(tweet_id)
    except Exception as e:
        print('error')
        error_list.append(tweet_id)

892420643555336193
892177421306343426
891815181378084864
891689557279858688
891327558926688256
891087950875897856
890971913173991426
890729181411237888
890609185150312448
890240255349198849
890006608113172480
889880896479866881
889665388333682689
889638837579907072
889531135344209921
889278841981685760
888917238123831296
888804989199671297
888554962724278272
error
888078434458587136
887705289381826560
887517139158093824
887473957103951883
887343217045368832
887101392804085760
886983233522544640
886736880519319552
886680336477933568
886366144734445568
886267009285017600
886258384151887873
886054160059072513
885984800019947520
885528943205470208
885518971528720385
885311592912609280
885167619883638784
884925521741709313
884876753390489601
884562892145688576
884441805382717440
884247878851493888
884162670584377345
883838122936631299
883482846933004288
883360690899218434
883117836046086144
882992080364220416
882762694511734784
882627270321602560
882268110199369728
882045870035918850
881906

819347104292290561
819238181065359361
819227688460238848
819015337530290176
819015331746349057
819006400881917954
819004803107983360
818646164899774465
818627210458333184
818614493328580609
818588835076603904
818536468981415936
818307523543449600
818259473185828864
818145370475810820
817908911860748288
817827839487737858
817777686764523521
817536400337801217
817502432452313088
817423860136083457
817415592588222464
817181837579653120
817171292965273600
817120970343411712
817056546584727552
816829038950027264
816816676327063552
816697700272001025
816450570814898180
816336735214911488
816091915477250048
816062466425819140
816014286006976512
815990720817401858
815966073409433600
815745968457060357
815736392542261248
815639385530101762
815390420867969024
814986499976527872
814638523311648768
814578408554463233
814530161257443328
814153002265309185
813944609378369540
813910438903693312
813812741911748608
813800681631023104
813217897535406080
813202720496779264
813187593374461952
813172488309

760656994973933572
760641137271070720
760539183865880579
760521673607086080
760290219849637889
760252756032651264
760190180481531904
error
759943073749200896
759923798737051648
759846353224826880
759793422261743616
759566828574212096
759557299618865152
759447681597108224
759446261539934208
759197388317847553
759159934323924993
759099523532779520
759047813560868866
758854675097526272
758828659922702336


Rate limit reached. Sleeping for: 689


758740312047005698
758474966123810816
758467244762497024
758405701903519748
758355060040593408
758099635764359168
758041019896193024
757741869644341248
757729163776290825
757725642876129280
757611664640446465
757597904299253760
757596066325864448
757400162377592832
757393109802180609
757354760399941633
756998049151549440
756939218950160384
756651752796094464
756526248105566208
756303284449767430
756288534030475264
756275833623502848
755955933503782912
755206590534418437
755110668769038337
754874841593970688
754856583969079297
754747087846248448
754482103782404096
754449512966619136
754120377874386944
error
753655901052166144
753420520834629632
753398408988139520
753375668877008896
753298634498793472
753294487569522689
753039830821511168
753026973505581056
752932432744185856
752917284578922496
752701944171524096
752682090207055872
752660715232722944
752568224206688256
752519690950500352
752334515931054080
752309394570878976
752173152931807232
751950017322246144
751937170840121344
751830

705475953783398401
705442520700944385
705428427625635840
705239209544720384
705223444686888960
705102439679201280
705066031337840642
704871453724954624
704859558691414016
704847917308362754
704819833553219584
704761120771465216
704499785726889984
704491224099647488
704480331685040129
704364645503647744
704347321748819968
704134088924532736
704113298707505153
704054845121142784
703774238772166656
703769065844768768
703631701117943808
703611486317502464
703425003149250560
703407252292673536
703382836347330562
703356393781329922
703268521220972544
703079050210877440
703041949650034688
702932127499816960
702899151802126337
702684942141153280
702671118226825216
702598099714314240
702539513671897089
702332542343577600
702321140488925184
702276748847800320
702217446468493312
701981390485725185
701952816642965504
701889187134500865
701805642395348998
701601587219795968
701570477911896070
701545186879471618
701214700881756160
700890391244103680
700864154249383937
700847567345688576
700796979434

678424312106393600
678410210315247616
678399652199309312
678396796259975168
678389028614488064
678380236862578688
678341075375947776
678334497360859136
678278586130948096
678255464182861824
678023323247357953
678021115718029313
677961670166224897
677918531514703872
677895101218201600
677716515794329600
677700003327029250
677698403548192770
677687604918272002
677673981332312066
677662372920729601
677644091929329666
677573743309385728
677565715327688705
677557565589463040
677547928504967168
677530072887205888
677335745548390400
677334615166730240
677331501395156992
677328882937298944
677314812125323265
677301033169788928
677269281705472000
677228873407442944


Rate limit reached. Sleeping for: 688


677187300187611136
676975532580409345
676957860086095872
676949632774234114
676948236477857792
676946864479084545
676942428000112642
676936541936185344
676916996760600576
676897532954456065
676864501615042560
676821958043033607
676819651066732545
676811746707918848
676776431406465024
676617503762681856
676613908052996102
676606785097199616
676603393314578432
676593408224403456
676590572941893632
676588346097852417
676582956622721024
676575501977128964
676533798876651520
676496375194980353
676470639084101634
676440007570247681
676430933382295552
676263575653122048
676237365392908289
676219687039057920
676215927814406144
676191832485810177
676146341966438401
676121918416756736
676101918813499392
676098748976615425
676089483918516224
675898130735476737
675891555769696257
675888385639251968
675878199931371520
675870721063669760
675853064436391936
675849018447167488
675845657354215424
675822767435051008
675820929667219457
675798442703122432
675781562965868544
675740360753160193
675710890956

668226093875376128
668221241640230912
668204964695683073
668190681446379520
668171859951755264
668154635664932864
668142349051129856
668113020489474048
667937095915278337
667924896115245057
667915453470232577
667911425562669056
667902449697558528
667886921285246976
667885044254572545
667878741721415682
667873844930215936
667866724293877760
667861340749471744
667832474953625600
667806454573760512
667801013445750784
667793409583771648
667782464991965184
667773195014021121
667766675769573376
667728196545200128
667724302356258817
667550904950915073
667550882905632768
667549055577362432
667546741521195010
667544320556335104
667538891197542400
667534815156183040
667530908589760512
667524857454854144
667517642048163840
667509364010450944
667502640335572993
667495797102141441
667491009379606528
667470559035432960
667455448082227200
667453023279554560
667443425659232256
667437278097252352
667435689202614272
667405339315146752
667393430834667520
667369227918143488
667211855547486208
667200525029

In [9]:
# create list to append information into
df3_list = []
# fill list with info from text files
for tweet_info in glob.glob('tweet_infos/*.txt'):
    with open(tweet_info, encoding='utf-8') as file:
        content = json.load(file)
        tweet_id = content['id']
        retweet_count = content['retweet_count']            
        favorite_count = content['favorite_count']
        df3_list.append({'tweet_id' : tweet_id,
                   'retweet_count' : retweet_count,
                   'favorite_count' : favorite_count})
        print(tweet_id)

666020888022790149
666029285002620928
666033412701032449
666044226329800704
666049248165822465
666050758794694657
666051853826850816
666055525042405380
666057090499244032
666058600524156928
666063827256086533
666071193221509120
666073100786774016
666082916733198337
666094000022159362
666099513787052032
666102155909144576
666104133288665088
666268910803644416
666273097616637952
666287406224695296
666293911632134144
666337882303524864
666345417576210432
666353288456101888
666362758909284353
666373753744588802
666396247373291520
666407126856765440
666411507551481857
666418789513326592
666421158376562688
666428276349472768
666430724426358785
666435652385423360
666437273139982337
666447344410484738
666454714377183233
666644823164719104
666649482315059201
666691418707132416
666701168228331520
666739327293083650
666776908487630848
666781792255496192
666786068205871104
666804364988780544
666817836334096384
666826780179869698
666835007768551424
666837028449972224
666983947667116034
666996132027

675820929667219457
675822767435051008
675845657354215424
675849018447167488
675853064436391936
675870721063669760
675878199931371520
675888385639251968
675891555769696257
675898130735476737
676089483918516224
676098748976615425
676101918813499392
676121918416756736
676146341966438401
676191832485810177
676215927814406144
676219687039057920
676237365392908289
676263575653122048
676430933382295552
676440007570247681
676470639084101634
676496375194980353
676533798876651520
676575501977128964
676582956622721024
676588346097852417
676590572941893632
676593408224403456
676603393314578432
676606785097199616
676613908052996102
676617503762681856
676776431406465024
676811746707918848
676819651066732545
676821958043033607
676864501615042560
676897532954456065
676916996760600576
676936541936185344
676942428000112642
676946864479084545
676948236477857792
676949632774234114
676957860086095872
676975532580409345
677187300187611136
677228873407442944
677269281705472000
677301033169788928
677314812125

711008018775851008
711306686208872448
711363825979756544
711652651650457602
711694788429553666
711732680602345472
711743778164514816
711968124745228288
711998809858043904
712065007010385924
712085617388212225
712092745624633345
712097430750289920
712309440758808576
712438159032893441
712668654853337088
712717840512598017
712809025985978368
713175907180089344
713177543487135744
713411074226274305
713761197720473600
713900603437621249
713909862279876608
713919462244790272
714141408463036416
714214115368108032
714251586676113411
714258258790387713
714485234495041536
714606013974974464
714631576617938945
714957620017307648
714962719905021952
714982300363173890
715009755312439296
715200624753819648
715220193576927233
715342466308784130
715360349751484417
715680795826982913
715696743237730304
715704790270025728
715733265223708672
715758151270801409
715928423106027520
716080869887381504
716285507865542656
716439118184652801
716447146686459905
716730379797970944
716791146589110272
716802964044

789960241177853952
789986466051088384
790227638568808452
790277117346975746
790337589677002753
790581949425475584
790698755171364864
790723298204217344
790946055508652032
790987426131050500
791026214425268224
791312159183634433
791406955684368384
791672322847637504
791774931465953280
791780927877898241
791784077045166082
791821351946420224
792050063153438720
792394556390137856
792773781206999040
792883833364439040
792913359805018113
793120401413079041
793135492858580992
793150605191548928
793165685325201412
793180763617361921
793195938047070209
793210959003287553
793226087023144960
793241302385262592
793256262322548741
793271401113350145
793286476301799424
793500921481273345
793601777308463104
793614319594401792
793845145112371200
793962221541933056
794205286408003585
794332329137291264
794355576146903043
794926597468000259
794983741416415232
795076730285391872
795400264262053889
795464331001561088
796031486298386433
796080075804475393
796116448414461957
796125600683540480
796149749086

863062471531167744
863079547188785154
863427515083354112
863432100342583297
863471782782697472
863553081350529029
863907417377173506
864197398364647424
864279568663928832
864873206498414592
865006731092295680
865359393868664832
865718153858494464
866094527597207552
866334964761202691
866450705531457537
866686824827068416
866720684873056260
867051520902168576
867072653475098625
867421006826221569
867774946302451713
867900495410671616
868552278524837888
868622495443632128
868639477480148993
868880397819494401
869227993411051520
869596645499047938
869702957897576449
869772420881756160
870063196459192321
870308999962521604
870374049280663552
870656317836468226
870726314365509632
870804317367881728
871032628920680449
871102520638267392
871166179821445120
871515927908634625
871762521631449091
871879754684805121
872122724285648897
872486979161796608
872620804844003328
872820683541237760
872967104147763200
873213775632977920
873337748698140672
873580283840344065
874012996292530176
874057562936

In [10]:
# convert list to DataFrame
df3 = pd.DataFrame(df3_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

### Assessing

In [11]:
df1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [12]:
df1.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [13]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

`df1` columns:
* **tweet_id**: the 18 digit unique id for each tweet in the provided WeRateDogs twitter archive
* **in_reply_to_status_id**: if a reply tweet, the 21 digit status id of the original tweet
* **in_reply_to_user_id**: if a reply tweet, the user id of the original tweet
* **timestamp**: the time the tweet was recorded
* **source**: the reference website the tweet was sent from (Twitter for iPhone, Twitter Web Client, Vine, or TweetDeck)
* **text**: the full text of the tweet including shortened link
* **retweeted_status_id**: if a retweet, the 21 digit status id of the original tweet
* **retweeted_status_user_id**: if a retweet, the user id of the original tweet
* **retweeted_status_timestamp**: if a retweet, the timestamp of the original tweet
* **expanded_urls**: the full url of the tweet
* **rating_numerator**: the numerator for the rating extracted from the text of the tweet
* **rating_denominator**: the denominator of the rating extracted from the text of the tweet
* **name**: the name of the dog extracted from the text of the tweet
* **doggo**: either 'none' or 'doggo' if 'doggo' is used in the text of the tweet
* **floofer**:either 'none' or 'floofer' if 'floofer' is used in the text of the tweet
* **pupper**:either 'none' or 'pupper' if 'pupper' is used in the text of the tweet
* **puppo**:either 'none' or 'puppo' if 'puppo' is used in the text of the tweet

In [14]:
df2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [15]:
df2.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [16]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


`df2` columns:
* **tweet_id**: the 18 digit unique id for each tweet in the provided WeRateDogs twitter archive
* **jpg_url**: the url of the image for the tweet
* **img_num**: 1-4 value???
* **p1**: the top dog breed prediction from the neural network
* **p1_conf**: the confidence of the prediction
* **p1_dog**: whether the prediction is a breed of dog
* **p2**: the second dog breed prediction from the neural network
* **p2_conf**: the confidence of the prediction
* **p2_dog**: whether the prediction is a breed of dog
* **p3**: the third dog breed prediction from the neural network
* **p3_conf**: the confidence of the prediction
* **p3_dog**: whether the prediction is a breed of dog|

In [17]:
df3.head()

tweet_id  retweet_count  favorite_count
0  666020888022790149            478            2478
1  666029285002620928             45             123
2  666033412701032449             43             119
3  666044226329800704            135             280
4  666049248165822465             41             101

In [18]:
df3.describe()

tweet_id  retweet_count  favorite_count
count  2.333000e+03    2333.000000     2333.000000
mean   7.419279e+17    2784.985426     7727.835405
std    6.820607e+16    4713.944624    11996.042902
min    6.660209e+17       1.000000        0.000000
25%    6.782786e+17     560.000000     1340.000000
50%    7.184547e+17    1302.000000     3367.000000
75%    7.986654e+17    3239.000000     9488.000000
max    8.924206e+17   80032.000000   160093.000000

In [19]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2333 entries, 0 to 2332
Data columns (total 3 columns):
tweet_id          2333 non-null int64
retweet_count     2333 non-null int64
favorite_count    2333 non-null int64
dtypes: int64(3)
memory usage: 54.8 KB


`df3` columns:
* **tweet_id**: the 18 digit unique id for each tweet in the provided WeRateDogs twitter archive
* **retweet_count**: the number of retweets the tweet received
* **favorite_count**: the number of favorites (likes) the tweet received

#### Quality
1. the df1 rating_numerator max values of 1776, 420, and 666 are specific joke values and so should be excluded from calculations
2. the df1 rating_numerator minimum value is 0 in two instances, which are both accurate to their tweets but are special cases which shouldn't be included in the rest of the data
3. for df1, the rating_numerator and rating_denominator for tweet_id 835246439529840640 should be 13 and 10, respectively.
4. in df1, some of the names are inputted incorrect and are random words from the text which should be fixed
5. all ID values should be stored as strings not floats 
6. not all doggo, floofer, pupper, and puppo are labelled correctly!!! 
7. in df2, should have a column which has the top prediction that is a type of dog and a corresponding column for the confidence of that prediction
8. calculations would be made easier by including a column of rating_numerators over rating_denominators

#### Tidiness
1. need to make the doggo, floofer, pupper, and puppo columns all one column
2. can actually put all data into df1 without sacrificing tidiness standards

In [ ]:
df1.tweet_id[0]

In [ ]:
df1.expanded_urls[0]

In [ ]:
int(df1.expanded_urls[0].split('/')[-3])

In [38]:
list1 = df1[df1['doggo'] == 'doggo'].tweet_id

In [40]:
df1[df1['text'].str.contains('doggo')]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
9     890240255349198849                    NaN                  NaN   
43    884162670584377345                    NaN                  NaN   
83    876537666061221889                    NaN                  NaN   
99    872967104147763200                    NaN                  NaN   
108   871515927908634625                    NaN                  NaN   
110   871102520638267392                    NaN                  NaN   
121   869596645499047938                    NaN                  NaN   
172   858843525470990336                    NaN                  NaN   
191   855851453814013952                    NaN                  NaN   
200   854010172552949760                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
240   846514051647705089                    NaN                  NaN   
248   845397057150107648                    NaN                  NaN   
268   841439858740625411                    NaN                  NaN   
300   836753516572119041                    NaN                  NaN   
302   836648853927522308                    NaN                  NaN   
318   834574053763584002                    NaN                  NaN   
331   832998151111966721                    NaN                  NaN   
339   832273440279240704                    NaN                  NaN   
344   832032802820481025                    NaN                  NaN   
345   831939777352105988                    NaN                  NaN   
351   831322785565769729                    NaN                  NaN   
359   829878982036299777                    NaN                  NaN   
362   829449946868879360                    NaN                  NaN   
363   829374341691346946                    NaN                  NaN   
372   828381636999917570                    NaN                  NaN   
384   826958653328592898                    NaN                  NaN   
385   826848821049180160                    NaN                  NaN   
389   826476773533745153                    NaN                  NaN   
391   826204788643753985                    NaN                  NaN   
...                  ...                    ...                  ...   
881   760521673607086080                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
899   758828659922702336                    NaN                  NaN   
914   757393109802180609                    NaN                  NaN   
919   756526248105566208                    NaN                  NaN   
924   755206590534418437                    NaN                  NaN   
934   753420520834629632                    NaN                  NaN   
944   752682090207055872                    NaN                  NaN   
946   752568224206688256                    NaN                  NaN   
948   752334515931054080                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
967   750381685133418496           7.501805e+17         4.717297e+09   
977   750011400160841729                    NaN                  NaN   
985   749075273010798592                    NaN                  NaN   
987   749036806121881602                    NaN                  NaN   
992   748692773788876800                    NaN                  NaN   
1030  745433870967832576                    NaN                  NaN   
1039  744234799360020481                    NaN                  NaN   
1051  742534281772302336                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1075  739623569819336705                    NaN                  NaN   
1079  739238157791694849                    NaN                  NaN   
1103  735256018284875776                    NaN                  NaN   
1113  733109485275860992                    NaN         

In [45]:
pd.set_option('display.max_colwidth', -1)
df1[df1['tweet_id'] == 876537666061221889]['text']

83    I can say with the pupmost confidence that the doggos who assisted with this search are heroic as h*ck. 14/10 for all https://t.co/8yoc1CNTsu
Name: text, dtype: object

In [146]:
df1[df1['rating_numerator'] == 0]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
315   835152434251116546 NaN                    NaN                    
1016  746906459439529985  7.468859e+17           4.196984e+09          

                      timestamp  \
315   2017-02-24 15:40:31 +0000   
1016  2016-06-26 03:22:31 +0000   

                                                                                  source  \
315   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1016  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                         text  \
315   When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag         
1016  PUPDATE: can't see any. Even if I could, I couldn't reach them to pet. 0/10 much disappointment https://t.co/c7WXaB2nqX   

      retweeted_status_id  retweeted_status_user_id  \
315  NaN                  NaN                         
1016 NaN                  NaN                         

     retweeted_status_timestamp  \
315   NaN                         
1016  NaN                         

                                                                                                                                                                                        expanded_urls  \
315   https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1   
1016  https://twitter.com/dog_rates/status/746906459439529985/photo/1                                                                                                                                   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
315   0                 10                  None  None  None    None   None  
1016  0                 10                  None  None  None    None   None

In [56]:
api.get_status(835246439529840640).text

"@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"

In [128]:
pd.set_option("display.max_rows", 99999)
df1['name'].value_counts()

None              745
a                 55 
Charlie           12 
Lucy              11 
Cooper            11 
Oliver            11 
Penny             10 
Tucker            10 
Lola              10 
Bo                9  
Winston           9  
Sadie             8  
the               8  
Bailey            7  
Buddy             7  
an                7  
Toby              7  
Daisy             7  
Koda              6  
Milo              6  
Dave              6  
Rusty             6  
Oscar             6  
Scout             6  
Bella             6  
Leo               6  
Jax               6  
Stanley           6  
Jack              6  
Finn              5  
Sammy             5  
Gus               5  
Oakley            5  
Alfie             5  
very              5  
Louis             5  
Sunny             5  
Chester           5  
Phil              5  
Bentley           5  
George            5  
Larry             5  
Chip              4  
Clarence          4  
Luna              4  
Duke      

In [145]:
not_name_id = []
pd.set_option('display.max_colwidth', -1)
for index, row in df1.iterrows():
    if row['name'].islower() == True:
        not_name.append(row['tweet_id'])
        print(row['tweet_id'])
        print(row['text'])

887517139158093824
I've yet to rate a Venezuelan Hover Wiener. This is such an honor. 14/10 paw-inspiring af (IG: roxy.thedoxy) https://t.co/20VrLAA8ba
881536004380872706
Here is a pupper approaching maximum borkdrive. Zooming at never before seen speeds. 14/10 paw-inspiring af 
(IG: puffie_the_chow) https://t.co/ghXBIIeQZF
869988702071779329
RT @dog_rates: We only rate dogs. This is quite clearly a smol broken polar bear. We'd appreciate if you only send dogs. Thank you... 12/10…
859196978902773760
We only rate dogs. This is quite clearly a smol broken polar bear. We'd appreciate if you only send dogs. Thank you... 12/10 https://t.co/g2nSyGenG9
855459453768019968
Guys, we only rate dogs. This is quite clearly a bulbasaur. Please only send dogs. Thank you... 12/10 human used pet, it's super effective https://t.co/Xc7uj1C64x
832645525019123713
There's going to be a dog terminal at JFK Airport. This is not a drill. 10/10  
https://t.co/dp5h9bCwU7
828650029636317184
Occasionally, we're se

### Cleaning

## Store

<a id='analyze'></a>
## Data Analysis

<a id='visualize'></a>
## Data Visualization

<a id='conclusion'></a>
## Conclusion

### Data Wrangling Effort

### Data Analyses and Visualizations